File: GitToolBox/src/test/java/zielu/gittoolbox/startup/ConfigMigratorV1toV2Test.java
package zielu.gittoolbox.startup;

import static org.assertj.core.api.Assertions.assertThat;
import static zielu.gittoolbox.config.DecorationPartType.BRANCH;
import static zielu.gittoolbox.config.DecorationPartType.LOCATION;
import static zielu.gittoolbox.config.DecorationPartType.STATUS;
import static zielu.gittoolbox.config.DecorationPartType.TAGS_ON_HEAD;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import zielu.gittoolbox.config.DecorationPartType;
import zielu.gittoolbox.config.GitToolBoxConfig;
import zielu.gittoolbox.config.GitToolBoxConfig2;

@Tag("fast")
class ConfigMigratorV1toV2Test {
  private ConfigMigratorV1toV2 migrator = new ConfigMigratorV1toV2();

  @Test
  void shouldMigrateLocationBeforeStatus() {
    GitToolBoxConfig v1 = new GitToolBoxConfig();
    GitToolBoxConfig2 v2 = migrate(v1);

    assertDecorationParts(v2, LOCATION, BRANCH, STATUS, TAGS_ON_HEAD);
  }

  @Test
  void shouldMigrateStatusBeforeLocation() {
    GitToolBoxConfig v1 = new GitToolBoxConfig();
    v1.showProjectViewStatusBeforeLocation = true;
    GitToolBoxConfig2 v2 = migrate(v1);

    assertDecorationParts(v2, BRANCH, STATUS, TAGS_ON_HEAD, LOCATION);
  }

  @Test
  void shouldMigrateShowTagsOnHeadDisabled() {
    GitToolBoxConfig v1 = new GitToolBoxConfig();
    v1.showProjectViewHeadTags = false;
    GitToolBoxConfig2 v2 = migrate(v1);

    assertDecorationParts(v2, LOCATION, BRANCH, STATUS);
  }

  @Test
  void shouldMigrateShowLocationPathDisabled() {
    GitToolBoxConfig v1 = new GitToolBoxConfig();
    v1.showProjectViewLocationPath = false;
    GitToolBoxConfig2 v2 = migrate(v1);

    assertDecorationParts(v2, BRANCH, STATUS, TAGS_ON_HEAD);
  }

  private GitToolBoxConfig2 migrate(GitToolBoxConfig v1) {
    GitToolBoxConfig2 v2 = new GitToolBoxConfig2();
    migrator.migrate(v1, v2);
    return v2;
  }

  private void assertDecorationParts(GitToolBoxConfig2 v2, DecorationPartType... types) {
    assertThat(v2.decorationParts).extracting(part -> part.type).containsExactly(types);
  }
}

File: GitToolBox/src/test/kotlin/zielu/gittoolbox/startup/ConfigMigratorV1toV2Test.kt
package zielu.gittoolbox.startup

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import zielu.gittoolbox.config.DecorationPartConfig
import zielu.gittoolbox.config.DecorationPartType
import zielu.gittoolbox.config.GitToolBoxConfig
import zielu.gittoolbox.config.GitToolBoxConfig2

internal class ConfigMigratorV1toV2Test {
  @Test
  fun shouldMigrateLocationBeforeStatus() {
    // given
    val v1 = GitToolBoxConfig()

    // when
    val v2 = migrate(v1)

    // then
    assertDecorationParts(
      v2,
      DecorationPartType.LOCATION,
      DecorationPartType.BRANCH,
      DecorationPartType.STATUS,
      DecorationPartType.TAGS_ON_HEAD
    )
  }

  @Test
  fun shouldMigrateStatusBeforeLocation() {
    // given
    val v1 = GitToolBoxConfig()
    v1.showProjectViewStatusBeforeLocation = true

    // when
    val v2 = migrate(v1)

    // then
    assertDecorationParts(
      v2,
      DecorationPartType.BRANCH,
      DecorationPartType.STATUS,
      DecorationPartType.TAGS_ON_HEAD,
      DecorationPartType.LOCATION
    )
  }

  @Test
  fun shouldMigrateShowTagsOnHeadDisabled() {
    // given
    val v1 = GitToolBoxConfig()
    v1.showProjectViewHeadTags = false

    // when
    val v2 = migrate(v1)

    // then
    assertDecorationParts(
      v2,
      DecorationPartType.LOCATION,
      DecorationPartType.BRANCH,
      DecorationPartType.STATUS
    )
  }

  @Test
  fun shouldMigrateShowLocationPathDisabled() {
    // given
    val v1 = GitToolBoxConfig()
    v1.showProjectViewLocationPath = false

    // when
    val v2 = migrate(v1)

    // then
    assertDecorationParts(
      v2,
      DecorationPartType.BRANCH,
      DecorationPartType.STATUS,
      DecorationPartType.TAGS_ON_HEAD
    )
  }

  private fun migrate(v1: GitToolBoxConfig): GitToolBoxConfig2 {
    val v2 = GitToolBoxConfig2()
    val migrator = ConfigMigratorV1toV2(v1)
    migrator.migrate(v2)
    return v2
  }

  private fun assertDecorationParts(v2: GitToolBoxConfig2, vararg types: DecorationPartType) {
    assertThat(v2.decorationParts)
      .extracting<DecorationPartType, RuntimeException> { part: DecorationPartConfig -> part.type }
      .containsExactly(*types)
  }
}
